File: test/kotlin/kotlinc/http-service/success-runtime-config/src/acceptance/http/HttpServer.kt
package acceptance.http
import java.io.IOException
import java.net.ServerSocket
class HttpServer(private val port: Int, private val router: Router) {
    fun run() {
        ServerSocket(port).use { server ->
            while (true) server.accept().use { socket ->
                try {
                    socket.soTimeout = 3000
                    val input = socket.getInputStream().bufferedReader(Charsets.US_ASCII)
                    val line = input.readLine() ?: return@use
                    var headerBytes = line.length
                    while (true) {
                        val header = input.readLine() ?: break
                        if (header.isEmpty()) break
                        headerBytes += header.length
                        if (headerBytes > 8192) throw IOException("request headers too large")
                    }
                    val target = line.split(' ').getOrElse(1) { "/" }
                    val response = router.route(target)
                    val body = response.body.toByteArray(Charsets.UTF_8)
                    val reason = if (response.status == 200) "OK" else if (response.status == 503) "Service Unavailable" else "Bad Request"
                    val headers = "HTTP/1.1 ${response.status} $reason\r\nContent-Type: ${response.type}\r\nContent-Length: ${body.size}\r\nConnection: close\r\n\r\n"
                    socket.getOutputStream().apply {
                        write(headers.toByteArray(Charsets.US_ASCII)); write(body); flush()
                    }
                } catch (error: IOException) { /* Disconnected probes do not stop the service. */ }
            }
        }
    }
}

File: test/single-language/ruby/rubycli/http-service/failure-health-rollback/lib/http_server.rb
require 'socket'
module Fixture
  class HttpServer
    def initialize(configuration, router)
      @configuration, @router = configuration, router
    end
    def run
      server = TCPServer.new('0.0.0.0', @configuration.port)
      loop do
        client = server.accept
        begin
          request = client.gets || ''
          while (line = client.gets) && line != "\r\n"; end
          target = request.split(' ')[1] || '/'
          path, query = target.split('?', 2)
          status, headers, chunks = @router.route(path, query)
          reason = status == 200 ? 'OK' : status == 503 ? 'Service Unavailable' : 'Bad Request'
          client.write("HTTP/1.1 #{status} #{reason}\r\n")
          headers.each { |key, value| client.write("#{key}: #{value}\r\n") }
          client.write("Connection: close\r\n\r\n")
          chunks.each { |chunk| client.write(chunk) }
        rescue IOError, SystemCallError
          # A disconnected probe does not stop the service.
        ensure
          client.close
        end
      end
    ensure
      server&.close
    end
  end
end
